client/hwlib/src/collectors/os_info.rs

```diff
@@ -68,47 +81,90 @@ impl TryFrom<&Path> for KernelPackage {
     }
 }
 
-pub(crate) fn get_architecture() -> Result<String> {
-    let arch = Command::new("dpkg")
-        .arg("--print-architecture")
-        .output()?
-        .stdout;
-    Ok(String::from_utf8(arch)?.trim().to_string())
+pub(crate) fn get_architecture(runner: &dyn CommandRunner) -> Result<String> {
+    let arch = runner.run_command("dpkg", &["--print-architecture"])?;
+    Ok(arch.trim().to_string())
 }
 
-pub(super) fn get_codename() -> Result<String> {
+pub(super) fn get_codename(runner: &dyn CommandRunner) -> Result<String> {
     lazy_static! {
         static ref CODENAME_RE: Regex = Regex::new(r"Codename:\s*(\S+)").unwrap();
     }
-    get_lsb_release_info("-c", &CODENAME_RE)
+    get_lsb_release_info("-c", &CODENAME_RE, runner)
 }
 
-pub(super) fn get_distributor() -> Result<String> {
+pub(super) fn get_distributor(runner: &dyn CommandRunner) -> Result<String> {
     lazy_static! {
         static ref DISTRIBUTOR_RE: Regex = Regex::new(r"Distributor ID:\s*(\S+)").unwrap();
     }
-    get_lsb_release_info("-i", &DISTRIBUTOR_RE)
+    get_lsb_release_info("-i", &DISTRIBUTOR_RE, runner)
 }
 
-pub(super) fn get_version() -> Result<String> {
+pub(super) fn get_version(runner: &dyn CommandRunner) -> Result<String> {
     lazy_static! {
         static ref VERSION_RE: Regex = Regex::new(r"Release:\s*(\S+)").unwrap();
     }
-    get_lsb_release_info("-r", &VERSION_RE)
+    get_lsb_release_info("-r", &VERSION_RE, runner)
 }
 
-fn get_lsb_release_info(flag: &str, re: &Regex) -> Result<String> {
-    let lsb_release_output = Command::new("lsb_release")
-        .arg(flag)
-        .output()
-        .context(format!(
-            "failed to execute lsb_release command with flag {}",
-            flag
-        ))?;
-    let output_str = String::from_utf8(lsb_release_output.stdout)
-        .context("failed to convert lsb_release output to UTF-8 string")?;
-    re.captures(&output_str)
+fn get_lsb_release_info(flag: &str, re: &Regex, runner: &dyn CommandRunner) -> Result<String> {
+    let lsb_release_output = runner.run_command("lsb_release", &[flag])?;
+    re.captures(&lsb_release_output)
         .and_then(|caps| caps.get(1))
         .map(|m| m.as_str().to_string())
         .ok_or_else(|| anyhow!("failed to capture information using regex"))
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::helpers::test_utils::get_test_filepath;
+    use mockall::mock;
+
+    mock! {
+        CommandRunner {}
+        impl CommandRunner for CommandRunner {
+            fn run_command<'a>(&self, cmd: &str, args: &[&'a str]) -> Result<String>;
+        }
+    }
```

Comment:
Rather than bringing in a mocking crate for this I'd be tempted to just hand write the impl you need for your test cases as the logic is fairly simple:

```rust
struct MockRunner<'a> {
    calls: HashMap<(&'a str, Vec<&'a str>), Result<&'a str>>,
}

impl<'a> MockRunner<'a> {
    fn new(calls: Vec<((&'a str, &[&'a str]), Result<&'a str>)>) -> Self {
        let calls = calls
            .into_iter()
            .map(|((cmd, args), res)| ((cmd, args.to_vec()), res))
            .collect();

        Self { calls }
    }
}

impl<'a> CommandRunner for MockRunner<'a> {
    fn run_command(&self, cmd: &str, args: &[&str]) -> Result<String> {
        match self.calls.get(&(cmd, args.to_vec())) {
            Some(res) => res.clone().map(String::from),
            None => panic!("missing mock: cmd={cmd:?} args={args:?}"),
        }
    }
}
```

I've put this in a [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c24709d65ab934711d094df1378362d4) for you to take a look at if you want to try it out :slightly_smiling_face:

Reply:
@TheSignPainter98 I've noticed when I use `vec`, I can specify and empty array, like `"lsmod", vec![]` and it works
When I replace the implementation with `type CommandKey<'args> = (&'args str, &'args [&'args str]);`, I get the following error:

```rust
("lsmod", &[]), Ok("Module Size Used\nsnd 61440 1\n")),
    |                        ^^^ expected an array with a fixed size of 1 element, found one with 0 elements
```

I don't feel like specifying `"lsmod", &[""]` is a good solution. What do you think?

Reply:
Erm, that's very strange... is there a reference-to-array type somewhere (`&[T; _]` as opposed to a slice `&[T]`) which is forcing this? I think using a vec is okay here, as although it may allocate more, these are just tests, so it's okay 

This actually just gave me an idea (NB: optional, the existing solution is fine), I think it's possible to simplify the tests a little by making the keys to the map be just `Vec<&str>`/`&[&str]` rather than `(&str, Vec<&str>)`/`(&str, &[&str])` as although the latter makes it impossible to write a test without a command, if someone's done that, the test is broken anyway, so it's okay to panic. I think this would make specifying the input to the mock visually cleaner